File: 1-100/41/41.py
def gen_pandigital(digits=9):
	"""generates pandigital numbers starting with the biggest"""
	
	def generator(start, digits_left):
		if len(digits_left) == 1:
			yield int(start + digits_left[0])
		else:
			for i in digits_left:
				for p in generator(start + i, filter(lambda x: x != i, digits_left)):
					yield p

	digits = [str(i) for i in range(digits,0,-1)]
	for i in generator("",digits):
		yield i

def prime_check(maxnum):
	"""Returns a function that checks if a number < maxnum is prime"""
	bucket = range(int(math.sqrt(maxnum)) + 1)
	bucket[1] = 0
	for i in bucket:
		if i <> 0:
			for not_prime in xrange(2*i, int(math.sqrt(maxnum)) +1, i):
				bucket[not_prime] = 0
	bucket = filter(lambda x: x <> 0, bucket)
	
	def isPrime(num):
		for i in bucket:
			if num % i == 0:
				return False
			if i > math.sqrt(num):
				return True
		return True

	return isPrime

def problem_41():
	isPrime = prime_check(7654321)
	for i in gen_pandigital(7):
		if isPrime(i):
			return i

File: 1-100/41/41.cpp
#include <iostream>

using namespace std;

bool IsPrime(int n)
{
	if (n % 2 == 0)
		return (n == 2);

	for (int d = 3; d * d < n; d += 2)
		if (n % d == 0)
			return false;

	return true;
}

bool Perm(int num, bool digitUsed[], int stage)
{
     if (stage == 7)
     {
        if (IsPrime(num))
        {
           cout << "FOUND " << num << endl;
           return true;
        }
     }
     else
     {
         for (int c = 7; c > 0; c--)
         {       
             if (!digitUsed[c])
             {
                 if (!( (stage == 6) && ( (c == 2) || (c == 4) || (c == 5) || (c == 6) ) ))
                 {
                     digitUsed[c] = true;
                     
                     if (Perm(num * 10 + c, digitUsed, stage + 1))
                         return true;
                        
                     digitUsed[c] = false;
                 }
             }
         }
     }
         
     return false;
}

void Problem41()
{
    bool digitUsed[8];
    for (int i = 0; i < 8; i++)
        digitUsed[i] = false;
        
    Perm(0, digitUsed, 0);
}
